### PR TITLE
add mem optimized grouped glu

### PR DESCRIPTION
**Pull request by vchiley**

### CODE REVIEW

File: megablocks/layers/glu.py [hunk 1]
```diff
@@ -42,6 +44,139 @@ def forward(self, x, topo):
...
+        dx = ddsd_out
+        gg.backend.gmm(dsdd_out, w1, batch_sizes, c=dx)
+        dx += gg.backend.gmm(dv1_out, v1, batch_sizes)
+        return dx, dw1, dv1, dw2, None, None, None, None
+
+memory_optimized_grouped_glu = MemoryOptimizedGroupedGLU.apply
+
+
 class GroupedGLU(SparseGLU):
     def forward(self, x, tokens_per_expert):
```
Comment (tgale96) [inline megablocks/layers/glu.py:181]: Noticing now - does this function need to scale_grad on the weights?

Comment (vchiley) [reply megablocks/layers/glu.py:181]: [here](https://github.com/stanford-futuredata/megablocks/pull/66/files#diff-6d3db0eaf79501bae35b15e2735c8cd0225650305971979eae5593c9a300ac30R183) as noted by Sasha

File: megablocks/layers/glu.py [hunk 2]
```diff
@@ -42,6 +44,139 @@ def forward(self, x, topo):
...
+
+        # Save input tensor, quantizing if needed
+        input_save_args = (x,)
+        if num_input_bits != -1:
+            x_q, x_scales = turbo.quantize_signed(x, num_bits=num_input_bits)
+            input_save_args = (x_q, x_scales)
+
+        # GeLU.
+        if num_remat_bits == -1:
+            def act_glu(x1, x2, activation_fn):
```
Comment (tgale96) [inline megablocks/layers/glu.py:70]: nit: do we need this fn? Seems easier to just write the body once?

File: megablocks/layers/glu.py [hunk 3]
```diff
@@ -42,6 +44,139 @@ def forward(self, x, topo):
...
+    @staticmethod
+    @torch.cuda.amp.custom_bwd
+    def backward(ctx, ddsd_out):
+        if (not ctx.needs_input_grad[0] or
+            not ctx.needs_input_grad[1] or
+            not ctx.needs_input_grad[2]):
+            raise ValueError("Expected all MLP inputs to need grad.")
+
+        # Unpack saved tensors; ugly because quantizing changes tensor count
+        #
```
Comment (tgale96) [inline megablocks/layers/glu.py:102]: rm blank comment line?

File: megablocks/layers/glu.py [hunk 4]
```diff
@@ -42,6 +44,139 @@ def forward(self, x, topo):
...
+            # sdd_out_q, sdd_out_scales, v1_out_q, v1_out_scales = saved_tensors[-4:]
+
+        # Rematerialize activation_fn output.
+        activation_fn = ctx.activation_fn
+        activation_grad_fn = None
+        if ctx.num_remat_bits == -1:
+            with torch.set_grad_enabled(True):
+                sdd_out.requires_grad = True
+                v1_out.requires_grad = True
+                def act_glu(x1, x2, activation_fn):
```
Comment (tgale96) [inline megablocks/layers/glu.py:128]: ditto above

File: megablocks/layers/mlp.py [hunk 5]
```diff
@@ -501,11 +501,12 @@ def backward(ctx, ddsd_out):
         #
         # NOTE: This reuses the dactivation_fn_out allocation.
         if ctx.num_remat_bits == -1:
-            if activation_grad_fn is not None:
```
Comment (tgale96) [inline megablocks/layers/mlp.py:504]: I think there are a bunch more of these in this file to update like this?

Comment (vchiley) [reply megablocks/layers/mlp.py:504]: I thought I got them all in the last PR, but I'll check again.

**Review by mvpatel2000 — APPROVED** [review]

**Pull request merged**